src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs

```diff
@@ -413,49 +426,148 @@ public static SequenceValidity PeekFirstSequence(ReadOnlySpan<byte> data, out in
                 }
-                consumed++;
+
+                ThrowHelper.ThrowArgumentException_InvalidUTF8(utf16Value.Slice(bytesWritten));
             }
-        }
 
-        private static void EscapeNextChars(ReadOnlySpan<char> value, int firstChar, Span<char> destination, ref int consumed, ref int written)
-        {
-            int nextChar = -1;
-            if (JsonHelpers.IsInRangeInclusive(firstChar, JsonConstants.HighSurrogateStartValue, JsonConstants.LowSurrogateEndValue))
+            Debug.Assert(toUtf8Status == OperationStatus.Done);
+            Debug.Assert(bytesConsumed == utf16Value.Length);
+
+            // Escape the bytes.
+            byte[] utf8ConvertedDestinationArray = null;
+            Span<byte> utf8ConvertedDestination;
+            length = checked(bytesWritten * JsonConstants.MaxExpansionFactorWhileEscaping);
+            if (length > JsonConstants.StackallocThreshold)
+            {
+                utf8ConvertedDestinationArray = ArrayPool<byte>.Shared.Rent(length);
+                utf8ConvertedDestination = utf8ConvertedDestinationArray;
+            }
+            else
             {
-                consumed++;
-                if (value.Length <= consumed || firstChar >= JsonConstants.LowSurrogateStartValue)
+                unsafe
                 {
-                    ThrowHelper.ThrowArgumentException_InvalidUTF16(firstChar);
+                    byte* ptr = stackalloc byte[JsonConstants.StackallocThreshold];
+                    utf8ConvertedDestination = new Span<byte>(ptr, JsonConstants.StackallocThreshold);
                 }
+            }
+
+            EscapeString(utf8Destination.Slice(0, bytesWritten), utf8ConvertedDestination, indexOfFirstByteToEscape: 0, encoder, out int convertedBytesWritten);
 
-                nextChar = value[consumed];
-                if (!JsonHelpers.IsInRangeInclusive(nextChar, JsonConstants.LowSurrogateStartValue, JsonConstants.LowSurrogateEndValue))
+            if (utf8DestinationArray != null)
+            {
+                utf8Destination.Slice(0, bytesWritten).Clear();
+                ArrayPool<byte>.Shared.Return(utf8DestinationArray);
+            }
+
+            // Convert byte to char.
+#if BUILDING_INBOX_LIBRARY
+            OperationStatus toUtf16Status = Utf8.ToUtf16(utf8ConvertedDestination.Slice(0, convertedBytesWritten), destination, out int bytesRead, out int charsWritten);
+            Debug.Assert(toUtf16Status == OperationStatus.Done);
+            Debug.Assert(bytesRead == convertedBytesWritten);
+#else
+            string utf16 = JsonReaderHelper.GetTextFromUtf8(utf8ConvertedDestination.Slice(0, convertedBytesWritten));
+            utf16.AsSpan().CopyTo(destination);
+            int charsWritten = utf16.Length;
+#endif
+            written += charsWritten;
+
+            if (utf8ConvertedDestinationArray != null)
+            {
+                utf8ConvertedDestination.Slice(0, written).Clear();
```

Comment:
For those who might come across this PR/change in the future, this should have been `convertedBytesWritten` and not `written`. Using `written` to slice could cause `ArugmentOutOfRangeException` in subtle test cases, based on certain strings where you had ascii/non-ascii text mixed causing `written + charsWritten` to go beyond the array-pool rented array (which is also a power of 2 which meant there would have been some extra space as well).
```suggestion
                utf8ConvertedDestination.Slice(0, convertedBytesWritten).Clear();
```

**Note:** What's in master and release/3.0 branches no longer use this code path and hence don't contain the bug (it uses the built-in JavascriptEncoder.Encode method).

Reply:
Thanks @chriscur for reporting the issue.